tests/functional/test_new_resolver_target.py

```diff
@@ -0,0 +1,74 @@
+import pytest
+
+from pip._internal.cli.status_codes import ERROR, SUCCESS
+from tests.lib.path import Path
+from tests.lib.wheel import make_wheel
+
+
+@pytest.fixture()
+def fake_wheel_tag():
+    return "fakepy1-fakeabi-fakeplat"
+
+
+@pytest.fixture()
+def fake_wheel(script, fake_wheel_tag):
+    wheel_house = script.scratch_path.joinpath("wheelhouse")
+    wheel_house.mkdir()
+    wheel_builder = make_wheel(
+        name="fake",
+        version="1.0",
+        wheel_metadata_updates={"Tag": []},
+    )
+    wheel_path = wheel_house.joinpath("fake-1.0-{}.whl".format(fake_wheel_tag))
+    wheel_builder.save_to(wheel_path)
+    return wheel_path
+
+
+@pytest.mark.parametrize("implementation", [None, "fakepy"])
+@pytest.mark.parametrize("python_version", [None, "1"])
+@pytest.mark.parametrize("abi", [None, "fakeabi"])
+@pytest.mark.parametrize("platform", [None, "fakeplat"])
+def test_new_resolver_target_checks_compatibility_failure(
+    script,
+    fake_wheel_tag,
+    fake_wheel,
+    implementation,
+    python_version,
+    abi,
+    platform,
+):
+    args = [
+        "install", "--unstable-feature=resolver",
+        "--only-binary=:all:",
+        "--no-cache-dir", "--no-index",
+        "--target", str(script.scratch_path.joinpath("target")),
+        fake_wheel,
+    ]
+    if implementation:
+        args += ["--implementation", implementation]
+    if python_version:
+        args += ["--python-version", python_version]
+    if abi:
+        args += ["--abi", abi]
+    if platform:
+        args += ["--platform", platform]
+
+    args_tag = "{}{}-{}-{}".format(
+        implementation,
+        python_version,
+        abi,
+        platform,
+    )
+
+    result = script.pip(
+        *args,
+        expect_error=(args_tag != fake_wheel_tag)
+    )
+
+    dist_info = Path("scratch", "target", "fake-1.0.dist-info")
+    if args_tag == fake_wheel_tag:
+        result.returncode == SUCCESS
```

Comment:
This should be an `assert`? Alternatively we can drop this check because it should already be asserted within `script.pip` itself.